File: AgeOfArenaUnity/Assets/Scripts/MarketSystem.cs
using UnityEngine;
using System.Collections.Generic;

/// <summary>
/// Fluctuating-price resource exchange driven by the Market building. Every
/// Sell of a commodity drives its price down; every Buy drives it up. Prices
/// drift back toward the base rate (0.7 sell / 1.3 buy) over time. The spread
/// between sell and buy is always maintained so trading is never free.
/// <see cref="CommandSystem"/> calls Sell/Buy on hotkeys; <see cref="GameManager"/>
/// calls <see cref="Tick"/> each update.
/// </summary>
public static class MarketSystem
{
    public const int Batch = 100;           // units moved per trade

    const float BaseSell = 0.7f;            // base sell rate (commodity → gold ratio)
    const float BaseBuy  = 1.3f;            // base buy  rate
    const float PriceShift = 0.05f;         // rate change per trade
    const float DriftRate  = 0.002f;        // drift back to base per second
    const float MinSell    = 0.3f;
    const float MaxBuy     = 2.5f;

    static readonly float[] _sellRate = { BaseSell, BaseSell, BaseSell }; // Food/Wood/Stone
    static readonly float[] _buyRate  = { BaseBuy,  BaseBuy,  BaseBuy  };

    const float GuildsAdjust = 0.10f;  // MKTT: Guilds narrows the spread by this much each side

    /// <summary>Guilds (player Market tech) raises sell and lowers buy → narrower spread.</summary>
    static float GuildsAdj()
        => (GameManager.Instance?.tech?.HasGuilds ?? false) ? GuildsAdjust : 0f;

    public static int SellGold(ResourceKind k)  => Mathf.RoundToInt(Batch * (_sellRate[Idx(k)] + GuildsAdj()));
    public static int BuyCost(ResourceKind k)   => Mathf.RoundToInt(Batch * (_buyRate[Idx(k)] - GuildsAdj()));

    /// <summary>Drift prices back to base over time. Call once per frame from GameManager.</summary>
    public static void Tick(float dt)
    {
        for (int i = 0; i < 3; i++)
        {
            _sellRate[i] = Mathf.MoveTowards(_sellRate[i], BaseSell, DriftRate * dt);
            _buyRate[i]  = Mathf.MoveTowards(_buyRate[i],  BaseBuy,  DriftRate * dt);
        }
    }

    /// <summary>Sell a batch of <paramref name="kind"/> for gold. No-op if short.</summary>
    public static void Sell(ResourceManager rm, ResourceKind kind)
    {
        if (rm == null || kind == ResourceKind.Gold) return;
        if (rm.Get(kind) < Batch) return;
        rm.Gain(kind, -Batch);
        rm.Gain(ResourceKind.Gold, SellGold(kind));
        int i = Idx(kind);
        _sellRate[i] = Mathf.Max(MinSell, _sellRate[i] - PriceShift);
        _buyRate[i]  = Mathf.Max(_sellRate[i] + 0.2f, _buyRate[i] - PriceShift * 0.5f);
    }

    /// <summary>Buy a batch of <paramref name="kind"/> with gold. No-op if short.</summary>
    public static void Buy(ResourceManager rm, ResourceKind kind)
    {
        if (rm == null || kind == ResourceKind.Gold) return;
        if (rm.gold < BuyCost(kind)) return;
        rm.Gain(ResourceKind.Gold, -BuyCost(kind));
        rm.Gain(kind, Batch);
        int i = Idx(kind);
        _buyRate[i]  = Mathf.Min(MaxBuy, _buyRate[i] + PriceShift);
        _sellRate[i] = Mathf.Min(_buyRate[i] - 0.2f, _sellRate[i] + PriceShift * 0.5f);
    }

    /// <summary>Current prices for HUD display (sell/buy per 100 units).</summary>
    public static (int sell, int buy) Rates(ResourceKind kind)
        => (SellGold(kind), BuyCost(kind));

    static int Idx(ResourceKind k) => k switch
    {
        ResourceKind.Food  => 0,
        ResourceKind.Wood  => 1,
        ResourceKind.Stone => 2,
        _                  => 0,
    };
}

File: web/src/game/MarketSystem.ts
/**
 * MarketSystem.ts — port of MarketSystem.cs.
 * Fluctuating-price resource exchange. Sell/Buy shift prices; drift restores base.
 * Costs: BaseSell 0.7 / BaseBuy 1.3, batch=100 resources, PriceShift=0.05.
 */
import { ResourceKind } from "../core/GameTypes";
import type { ResourceManager } from "../core/ResourceManager";

const BaseSell   = 0.7;
const BaseBuy    = 1.3;
const PriceShift = 0.05;
const DriftRate  = 0.003;  // per second
const MinSell    = 0.3;
const MaxBuy     = 2.5;
const Batch      = 100;    // resources per transaction

/** Market prices are shared across all player instances (1v1). */
const _sellRate: number[] = [BaseSell, BaseSell, BaseSell]; // food, wood, stone
const _buyRate:  number[] = [BaseBuy,  BaseBuy,  BaseBuy];

function idx(kind: ResourceKind): number | null {
  switch (kind) {
    case ResourceKind.Food:  return 0;
    case ResourceKind.Wood:  return 1;
    case ResourceKind.Stone: return 2;
    default: return null; // Gold can't be traded for gold
  }
}

export class MarketSystem {
  /** Drift prices back toward base each tick. */
  tick(dt: number) {
    for (let i = 0; i < 3; i++) {
      _sellRate[i] = moveTowards(_sellRate[i], BaseSell, DriftRate * dt);
      _buyRate[i]  = moveTowards(_buyRate[i],  BaseBuy,  DriftRate * dt);
      // Maintain minimum spread
      if (_buyRate[i] < _sellRate[i] + 0.2) _buyRate[i] = _sellRate[i] + 0.2;
    }
  }

  /** Gold received for selling 100 of kind. */
  sellGold(kind: ResourceKind): number {
    const i = idx(kind);
    return i === null ? 0 : Math.round(Batch * _sellRate[i]);
  }

  /** Gold cost to buy 100 of kind. */
  buyCost(kind: ResourceKind): number {
    const i = idx(kind);
    return i === null ? 0 : Math.round(Batch * _buyRate[i]);
  }

  /** Sell 100 of kind for gold. Returns gold received, or 0 on failure. */
  sell(rm: ResourceManager, kind: ResourceKind): number {
    const i = idx(kind);
    if (i === null) return 0;
    if (rm.get(kind) < Batch) return 0;
    const gold = this.sellGold(kind);
    rm.deduct(
      kind === ResourceKind.Food  ? Batch : 0,
      kind === ResourceKind.Wood  ? Batch : 0,
      0,
      kind === ResourceKind.Stone ? Batch : 0,
    );
    rm.gain(ResourceKind.Gold, gold);
    _sellRate[i] = Math.max(MinSell, _sellRate[i] - PriceShift);
    _buyRate[i]  = Math.max(_sellRate[i] + 0.2, _buyRate[i] - PriceShift * 0.5);
    return gold;
  }

  /** Buy 100 of kind with gold. Returns true on success. */
  buy(rm: ResourceManager, kind: ResourceKind): boolean {
    const i = idx(kind);
    if (i === null) return false;
    const cost = this.buyCost(kind);
    if (!rm.canAfford(0, 0, cost)) return false;
    rm.deduct(0, 0, cost);
    rm.gain(kind, Batch);
    _buyRate[i]  = Math.min(MaxBuy, _buyRate[i]  + PriceShift);
    _sellRate[i] = Math.min(_buyRate[i] - 0.2, _sellRate[i] + PriceShift * 0.5);
    return true;
  }

  /** Reset prices (call on game start). */
  reset() {
    for (let i = 0; i < 3; i++) { _sellRate[i] = BaseSell; _buyRate[i] = BaseBuy; }
  }
}

function moveTowards(current: number, target: number, delta: number): number {
  return Math.abs(target - current) <= delta ? target
    : current + Math.sign(target - current) * delta;
}
